File: automation/cj_takt.mjs
// cj_takt.mjs — dieselbe Stempeluhr wie cj_takt.py: RESERVIERTE Startzeiten, Start-zu-Start ABSTAND,
// mkdir-Sperre und gemeinsame Stempeldatei, damit Python- und Node-Prozesse EINE Uhr teilen.
import fs from 'node:fs';
const LOCKDIR = '/tmp/cj_takt.lockdir', STEMPEL = '/tmp/cj_takt.stempel';
const ABSTAND = parseFloat(process.env.CJ_TAKT_S || '1.8');
const sleep = ms => new Promise(r => setTimeout(r, ms));
async function sperren() {
  for (;;) {
    try { fs.mkdirSync(LOCKDIR); return; } catch {}
    try { if (Date.now() - fs.statSync(LOCKDIR).mtimeMs > 5000) { fs.rmdirSync(LOCKDIR); continue; } } catch {}
    await sleep(20);
  }
}
// 24.09.2026 VORRANG (siehe cj_takt.py): solange cj_kosten_backfill arbeitet (frischt /tmp/cj_vorrang je Aufruf auf),
// warten alle ausser Bestell-/Zahlungs-/Versandwaechtern; Datei > 120 s = Leiche; hoechstens VORRANG_MAX_S warten.
const VORRANG_DATEI = '/tmp/cj_vorrang', VORRANG_SKRIPTE = ['cj_kosten_backfill'];
const IMMER_FREI = ['cj_fulfill', 'cj_order', 'cj_zahlung', 'versand_stillstand', 'bestell', 'cj_takt'];
const VORRANG_MAX_S = parseFloat(process.env.VORRANG_MAX_S || '5400');
async function vorrangWarten() {
  const name = (process.argv[1] || '?').split('/').pop();
  if (VORRANG_SKRIPTE.some(v => name.startsWith(v))) { try { fs.writeFileSync(VORRANG_DATEI, name); } catch {} return; }
  if (IMMER_FREI.some(v => name.startsWith(v))) return;
  // Globaler Vorrang (siehe cj_takt.py): Repo-Datei dropship/_cj_vorrang_global mit ISO-Enddatum → gemeinsamer Abstand je
  // Maschine (/tmp/cj_bremse_letzter, GLOBAL_ABSTAND_S = 180 s) für alle nicht dringenden Aufrufe.
  try {
    const gdat = new URL('../dropship/_cj_vorrang_global', import.meta.url);
    const bis = Date.parse(fs.readFileSync(gdat, 'utf8').trim().split(/\s+/)[0]);
    if (bis > Date.now()) {
      const ABSTAND = parseFloat(process.env.GLOBAL_ABSTAND_S || '180') * 1000, UHR = '/tmp/cj_bremse_letzter';
      for (;;) {
        let alter; try { alter = Date.now() - fs.statSync(UHR).mtimeMs; } catch { alter = ABSTAND; }
        if (alter >= ABSTAND) { try { fs.writeFileSync(UHR, name); } catch {} break; }
        await sleep(Math.min(ABSTAND - alter, 30000) + 100);
      }
    }
  } catch {}
  const ende = Date.now() + VORRANG_MAX_S * 1000;
  while (Date.now() < ende) {
    let alter; try { alter = Date.now() - fs.statSync(VORRANG_DATEI).mtimeMs; } catch { return; }
    if (alter > 120000) return;
    await sleep(10000);
  }
}
export async function takt(abstand = ABSTAND) {
  await vorrangWarten();
  await sperren();
  let start;
  try {
    let letzter = 0; try { letzter = parseFloat(fs.readFileSync(STEMPEL, 'utf8').split(/\s+/)[0]) || 0; } catch {}
    start = Math.max(Date.now() / 1000, letzter + abstand);
    fs.writeFileSync(STEMPEL, start.toFixed(3));
  } finally { try { fs.rmdirSync(LOCKDIR); } catch {} }
  // 24.09.2026: Protokoll je Aufruf (Epoche, Skript) — wer verbraucht das CJ-Tagesbudget? Nie den Aufruf stören.
  try {
    const d = new Date(start * 1000).toISOString().slice(0, 10);
    const name = (process.argv[1] || '?').split('/').pop();
    fs.appendFileSync(`/tmp/cj_takt_${d}.log`, `${Math.floor(start)}\t${name}\n`);
  } catch {}
  const warte = start * 1000 - Date.now();
  if (warte > 0) await sleep(warte);
}
export async function frei() { return null; }   // Kompatibilitaet, ohne Wirkung

File: automation/cj_takt.py
"""cj_takt.py — EIN Takt für ALLE CJ-Verbraucher (21.09.2026, «schneller automation»).

Gemessen: acht CJ-Aufrufe hintereinander OHNE Takt → jeder zweite 1600200 (QPS), obwohl kein
anderer Prozess lief; die Helfer antworten darauf mit 8/16/24 s Strafschlaf. So kostete der
Varianten-Wächter ~12 s je Produkt, obwohl CJ selbst in 0,6 s antwortet. Zweite Messung: ein
EINZELNER Prozess mit 1,05 s Start-zu-Start wird noch in 1 von 6 Fällen gedrosselt, mit 2,1 s
nie — CJ zählt offenbar ab ANTWORT-ENDE. Deshalb misst diese Uhr Ende-zu-Start.

Dritte Messung (Python 0/8, Node 8/8 gedrosselt): wer sich am Start-Stempel einreiht, kommt
0,4 s nach dem ENDE des anderen dran. Darum RESERVIERTE Startzeiten: `takt()` holt unter der
Sperre den nächsten freien Startpunkt (letzter reservierter Start + ABSTAND), trägt ihn ein,
gibt die Sperre frei und schläft bis dahin. Der Abstand ist Start-zu-Start und enthält die
Antwortdauer (~0,6 s) — 1,8 s lässt ≥0,9 s Ruhe zwischen Antwort-Ende und nächstem Start.
`frei()` bleibt als Aufrufkompatibilität, tut nichts mehr.
Sperre: atomares mkdir (dasselbe Protokoll wie cj_takt.mjs, damit Python- und Node-Prozesse
EINE Uhr teilen); eine Sperre älter als 5 s gilt als Leiche. Kein Prozess kann sie dauerhaft
blockieren.
"""
import os, time

LOCKDIR = "/tmp/cj_takt.lockdir"
STEMPEL = "/tmp/cj_takt.stempel"        # naechster reservierter Start (Epoche, s)
ABSTAND = float(os.environ.get("CJ_TAKT_S", "1.8"))


def _sperren():
    while True:
        try:
            os.mkdir(LOCKDIR); return
        except FileExistsError:
            try:
                if time.time() - os.stat(LOCKDIR).st_mtime > 5:
                    os.rmdir(LOCKDIR); continue
            except FileNotFoundError:
                continue
            time.sleep(0.02)


def _frei_sperre():
    try: os.rmdir(LOCKDIR)
    except FileNotFoundError: pass


def _lesen():
    try:
        return float(open(STEMPEL).read().split()[0])
    except Exception:
        return 0.0


def takt(abstand=None):
    """Reserviert den naechsten freien Startpunkt und blockiert bis dahin (Start-zu-Start ABSTAND)."""
    abstand = ABSTAND if abstand is None else abstand
    _sperren()
    try:
        start = max(time.time(), _lesen() + abstand)
        with open(STEMPEL, "w") as f:
            f.write(f"{start:.3f}")
    finally:
        _frei_sperre()
    _protokoll(start)
    warte = start - time.time()
    if warte > 0:
        time.sleep(warte)


def _protokoll(start):
    """24.09.2026: WER verbraucht das CJ-Tagesbudget? Je Aufruf eine Zeile (Epoche, Skript) in
    /tmp/cj_takt_<Datum>.log — sonst ist «der Kosten-Nachtrag bekommt nur 500 Produkte/Tag» nicht zu
    klären (30 Verbraucher, ein Budget). Fehler hier dürfen den Aufruf nie stören."""
    try:
        import sys
        name = os.path.basename(sys.argv[0] or "?") or "?"
        with open(time.strftime("/tmp/cj_takt_%Y-%m-%d.log", time.gmtime(start)), "a") as f:
            f.write(f"{int(start)}\t{name}\n")
    except Exception:
        pass


def frei():
    """Kompatibilitaet: frueher Ende-Stempel, seit der dritten Messung ohne Wirkung."""
    return None


if __name__ == "__main__":
    n = int(os.environ.get("N", "5")); t = []
    for _ in range(n):
        takt(); t.append(time.time())
    d = [round(b - a, 3) for a, b in zip(t, t[1:])]
    print("Start-Abstaende:", d, "| ok" if all(x >= ABSTAND - 0.02 for x in d) else "| ZU DICHT")
